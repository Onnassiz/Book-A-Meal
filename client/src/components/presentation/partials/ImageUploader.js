import React, { Component } from 'react';
import Dropzone from 'react-dropzone';
import axios from 'axios';
import PropTypes from 'prop-types';
import loaderImage from '../../../../assets/images/file-loader.gif';


class ImageUploader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showLoader: false,
      errors: {},
    };
    this.handleDrop = this.handleDrop.bind(this);
  }

  toggleLoader() {
    this.setState({ showLoader: !this.state.showLoader });
  }

  handleDrop(files) {
    const { putImage } = this.props;
    const url = 'https://api.cloudinary.com/v1_1/onnassiz/image/upload';
    const uploader = () => {
      const formData = new FormData();
      formData.append('file', files[0]);
      formData.append('tags', 'book-a-meal');
      formData.append('upload_preset', 'nbo5oyfm');
      formData.append('api_key', '258613626473737');
      formData.append('timestamp', (Date.now() / 1000));

      delete axios.defaults.headers.common.Authorization;
      this.toggleLoader();
      axios.post(url, formData, {
        headers: { 'X-Requested-With': 'XMLHttpRequest' },
      }).then((response) => {
        axios.defaults.headers.common.Authorization = `Bearer ${localStorage.getItem('token')}`;
        const { data } = response;
        const fileURL = data.secure_url; // You should store this URL for future references in your app
        this.toggleLoader();
        putImage(fileURL);
      }).catch((error) => {
        this.showErrors(error.response);
      });
    };
    uploader();
  }

  showErrors(errors) {
    this.setState({ errors });
  }

  render() {
    const dropZoneStyle = {
      width: 350,
      height: 300,
      margin: 'auto',
      padding: 20,
      border: 'solid #7a604a 1px',
      borderRadius: 3,
    };

    return (
			<div className="col-12">
				<div className="show-errors">
					<ul>
						{Object.keys(this.state.errors).map(item => <li key={item}>{ this.state.errors[item] }</li>)}
					</ul>
				</div>
				<Dropzone
					onDrop={this.handleDrop}
					accept="image/*"
          style={dropZoneStyle}
          className="dropzone"
				>
					<div style={{ textAlign: 'center' }}>
						<h2 id="dropzoneHeader">Drop Images Here</h2>
						<br /><br /><br />
						<button disabled={this.state.showLoader} className="button">Upload Image</button>
						<div>
							<img src={loaderImage} alt="loader" hidden={!this.state.showLoader} />
						</div>
					</div>
				</Dropzone>
			</div>
    );
  }
}

ImageUploader.propTypes = {
  putImage: PropTypes.func.isRequired,
};

export default ImageUploader;
