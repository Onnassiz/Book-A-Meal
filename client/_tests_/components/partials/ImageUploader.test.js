import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import LocalStorageMock from '../../mocks/localStorage';
import ImageUploader from '../../../src/components/presentation/partials/ImageUploader';

const props = {
  putImage: jest.fn(),
  isShowingAddPhoto: false,
  toggleAddPhoto: jest.fn(),
};

axios.defaults.headers.common.Authorization = 'Bearer Hello';
const mock = new MockAdapter(axios);
const url = 'https://api.cloudinary.com/v1_1/onnassiz/image/upload';

const wrapper = shallow(<ImageUploader {...props} />);

global.localStorage = new LocalStorageMock();

describe('<ImageUploader />', () => {
  it('Should render Image Uploader component successfully', () => {
    expect(wrapper.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should upload image successfully and return an image url', () => {
    const response = { secure_url: 'https://uploadedImage.jpeg' };
    localStorage.setItem('token', 'This is the token');

    mock.onPost(url).reply(200, response);

    const handleDropSpy = jest.spyOn(wrapper.instance(), 'handleDrop');

    wrapper.instance().handleDrop(['this is an image']).then((res) => {
      expect(res.data).toEqual(response);
      expect(wrapper.state().progress).toBe(0);
      expect(wrapper.state().showLoader).toBe(false);
      expect(handleDropSpy).toHaveBeenCalled();
    });
  });

  it('Should should return error if image upload fails', () => {
    const response = { message: 'An error occurred while uploading your image' };

    mock.onPost(url).reply(400, response);

    const handleDropSpy = jest.spyOn(wrapper.instance(), 'handleDrop');

    wrapper.instance().handleDrop(['this is an image']).then((err) => {
      expect(err.response.data).toEqual(response);
      expect(wrapper.state().errors).toEqual(response);
      expect(wrapper.state().showLoader).toBe(false);
      expect(handleDropSpy).toHaveBeenCalled();
    });
  });

  it('Should calculate upload progress when called', () => {
    const progress = { loaded: 3000, total: 5000 };

    const showProgressSpy = jest.spyOn(wrapper.instance(), 'showProgress');

    wrapper.instance().showProgress(progress);
    expect(wrapper.state().progress).toBe('60');
    expect(showProgressSpy).toHaveBeenCalled();
  });
});
