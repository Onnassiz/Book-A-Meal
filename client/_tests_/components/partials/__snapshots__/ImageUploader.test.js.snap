// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageUploader /> Should render Image Uploader component successfully 1`] = `
<div
  className="col-12"
>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    className="image-upload"
    closeTimeoutMS={1}
    contentLabel="Modal"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "WebkitOverflowScrolling": "touch",
          "background": "#ffffff",
          "border": "none",
          "borderRadius": "6px",
          "bottom": "30%",
          "left": "33%",
          "outline": "none",
          "overflow": "auto",
          "padding": "15px",
          "position": "absolute",
          "right": "33%",
          "top": "15%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.4)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
  >
    <div
      className="col-12"
    >
      <div
        className="col-12"
      >
        <a
          onClick={[MockFunction]}
          style={
            Object {
              "float": "right",
            }
          }
        >
          <i
            className="material-icons"
            style={
              Object {
                "fontSize": 25,
              }
            }
          >
            close
          </i>
        </a>
      </div>
      <div
        className="show-errors"
      >
        <ul />
      </div>
      <t
        accept="image/*"
        className="dropzone"
        disableClick={false}
        disablePreview={false}
        disabled={false}
        maxSize={Infinity}
        minSize={0}
        multiple={true}
        onDrop={[Function]}
        preventDropOnDocument={true}
        style={
          Object {
            "border": "solid #7a604a 1px",
            "borderRadius": 3,
            "height": 300,
            "margin": "auto",
            "marginTop": 70,
            "padding": 20,
            "width": 350,
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          <h2
            id="dropzoneHeader"
          >
            Drop Image Here
          </h2>
          <br />
          <br />
          <br />
          <button
            className="button"
            disabled={false}
            id="upload_image"
          >
            Upload Image
          </button>
          <div>
            <img
              alt="loader"
              hidden={true}
              src="https://res.cloudinary.com/onnassiz/image/upload/v1535468238/file-loader_msna34.gif"
            />
          </div>
          <div
            className="progress"
          />
        </div>
      </t>
    </div>
  </Modal>
</div>
`;
